### PR TITLE
make changepassword a controller

### LINKED ISSUE #15940: [enc2] changepassword controller breaks encryption modularity
https://github.com/owncloud/core/blob/master/settings/changepassword/controller.php#L82-L109 is making calls straight to default encryption app methods.

These need to be decoupled immediately as they completely defeat the purpose of what we did in the refactor.

Also the `changeUserPassword` method errors have the chance of falling through and showing multiple errors that the frontend can't handle.

Also also controller should be made a REAL controller to enable ability to remov eusage of depreciated methods such as `OC_JSON`

Cc @DeepDiver1975 @schiesbn @PVince81 

related to https://github.com/owncloud/core/issues/15823 and https://github.com/owncloud/core/issues/15822

### PR DESCRIPTION
**Pull request by th3fallen**
resolves https://github.com/owncloud/core/issues/15940

from team weak :)

@schiesbn per our conversations here we go (thanks for the suggestion)

Change to encryption:
- Default encryption module now uses its own recovery class (reason below)
- New method required by iEncryptionModule to implement the open/closed principle (SOLID) to the change password controller and make the code there module agnostic.

### CODE REVIEW

**Comment by th3fallen** [comment]
Squashed

**Comment by th3fallen** [comment]
> Test FAILed.

buh buh buh you didnt even give me a build link.... :frowning:

**Comment by Xenopathic** [comment]
@th3fallen https://ci.owncloud.org/job/pull-request-analyser-ng-simple/13499/ (click the Details link on the GitHub merge status bar). Looks like it was (another) Jenkins fall over.

File: lib/public/encryption/iencryptionmodule.php [hunk 1]
```diff
@@ -134,4 +134,14 @@ public function getUnencryptedBlockSize();
...
 	public function isReadable($path, $uid);
 
+	/**
+	 * Check if the recovery password presented is correct for the user
+	 *
+	 * @param $recoveryPassword
+	 * @param $user
+	 * @return array
+	 */
+	public function checkRecoveryPassword($recoveryPassword, $user);
```
Comment (schiesbn) [inline lib/public/encryption/iencryptionmodule.php]: Not sure if this should be in the module interface. Is this really something we need for every encryption module?
Does every encryption module will have a way to recover files on password lose?
Does every recovery procedure will be based on a (single) recovery key password?

The new changePassword controller is a great improvement! But maybe we should separate the recovery procedure completely from the change password dialog? Something like I described here: https://github.com/owncloud/core/issues/15940#issuecomment-97729211

**Comment by th3fallen** [comment]
@owncloud-bot retest this please

File: lib/public/encryption/iencryptionmodule.php (hunk 1 above)
Comment (th3fallen) [inline lib/public/encryption/iencryptionmodule.php]: I thought about that too, possibly moving it to its own interface that's optional (IchecksRecoveryPassword) or something of the likes, but then the issue comes in of how to pass that too the password controller and still make it module agnostic.

Thoughts?

File: apps/encryption/lib/crypto/encryption.php [hunk 2]
```diff
@@ -413,4 +432,26 @@ protected function getPathToRealFile($path) {
 		return $realPath;
 	}
 
+	/**
+	 * @param $recoveryPassword
+	 * @param $user
+	 * @return array
+	 */
+	public function checkRecoveryPassword($recoveryPassword, $user) {
```
Comment (rullzer) [inline apps/encryption/lib/crypto/encryption.php]: This function needs @since 8.2.0

Comment (nickvergessen) [inline apps/encryption/lib/crypto/encryption.php]: no, the public interface needs it ;)

File: lib/public/encryption/iencryptionmodule.php [hunk 3]
```diff
@@ -134,4 +134,14 @@ public function getUnencryptedBlockSize();
 	 */
 	public function isReadable($path, $uid);
 
+	/**
+	 * Check if the recovery password presented is correct for the user
+	 *
+	 * @param $recoveryPassword
+	 * @param $user
+	 * @return array
```
Comment (nickvergessen) [inline lib/public/encryption/iencryptionmodule.php]: This function needs `@since 8.2.0`

File: apps/encryption/lib/crypto/encryption.php (hunk 2 above)
Comment (rullzer) [inline apps/encryption/lib/crypto/encryption.php]: Ah indeed ;)

**Comment by PVince81** [comment]
What's the next step ?

**Comment by PVince81** [comment]
Anyone wants to pick this up ?

**Comment by LukasReschke** [comment]
Does not seem that way plus the code has already changed a bit => Let's close this.

**Pull request closed without merging**